Fix OSError message when importing CairoSvg

bonicim opened the pull request:
## What is the change?
* Create separate catch block for OSError when Cairo, a non-Python module, is not installed
* Update README with prerequisites and instructions on installing Cairo

Fixes #4.

## Why is it needed?
When `drawSvg` imports `cairosvg` in [this line of code](https://github.com/cduck/drawSvg/blob/bab21b78cc0e4c01ebb4371f24ab36781bc4e8d2/drawSvg/raster.py#L5), it currently catches both ImportError and OSError in the same block. However, if a user's platform does not have Cairo, a non-Python module, installed, it will throw a misleading error message, telling the user to install `cairosvg` even though it is already installed and hiding the real cause of the error, namely that `Cairo` is not installed. This change fixes that misleading error message. 

## Tests
Operating system: macOS Mojave (10.14.6)
Python Version: 3.7.4 (default, July 8, 2019)
I'm using a virtual environment.

* Cairo and CairoSvg not installed
```
In [1]: import drawSvg
/Users/markbonicillo/Workplace/drawSvg/drawSvg/raster.py:16: RuntimeWarning: CairoSVG will need to be installed to rasterize images: Install with `pip3 install cairosvg`
  warnings.warn(msg, RuntimeWarning)
```

* CairoSvg installed, but not Cairo
```
In [1]: import drawSvg
/Users/markbonicillo/Workplace/drawSvg/drawSvg/raster.py:12: RuntimeWarning: Cairo will need to be installed to rasterize images: See prerequisites section in `README`
  warnings.warn(msg, RuntimeWarning)
```

* CairoSvg and Cairo both installed
```
In [1]: import drawSvg as draw

d = draw.Drawing(200, 100, origin='center')

d.append(draw.Lines(-80, -45,
                    70, -49,
                    95, 49,
                    -90, 40,
                    close=False,
            fill='#eeee00',
            stroke='black'))

d.append(draw.Rectangle(0,0,40,50, fill='#1248ff'))
d.append(draw.Circle(-40, -10, 30,
            fill='red', stroke_width=2, stroke='black'))

p = draw.Path(stroke_width=2, stroke='green',
              fill='black', fill_opacity=0.5)
p.M(-30,5)  # Start path at point (-30, 5)
p.l(60,30)  # Draw line to (60, 30)
p.h(-70)    # Draw horizontal line to x=-70
p.Z()       # Draw line to start
d.append(p)

d.append(draw.ArcLine(60,-20,20,60,270,
            stroke='red', stroke_width=5, fill='red', fill_opacity=0.2))
d.append(draw.Arc(60,-20,20,60,270,cw=False,
            stroke='green', stroke_width=3, fill='none'))
d.append(draw.Arc(60,-20,20,270,60,cw=True,
            stroke='blue', stroke_width=1, fill='black', fill_opacity=0.3))

d.setPixelScale(2)  # Set number of pixels per geometry unit
#d.setRenderSize(400,200)  # Alternative to setPixelScale
d.saveSvg('example.svg')
d.savePng('example.png')

# Display in iPython notebook
d.rasterize()  # Display as PNG
d  # Display as SVG

Out[4]: <drawSvg.drawing.Drawing at 0x1177cad10>
```

cduck commented on README.md:
```diff
@@ -6,6 +6,23 @@ Most common SVG tags are supported and others can easily be added by writing a s
 An interactive [Jupyter notebook](https://jupyter.org) widget, `drawSvg.widgets.DrawingWidget`, is included that can update drawings based on mouse events.
 
+# Prerequisites
+Cairo needs to be installed separately. See platform-specific [instructions for Linux, Windows, and macOS from Cairo.](https://www.cairographics.org/download/) Below are some examples for installing Cairo on Linux distributions and macOS.
```
Add a sentence explaining what Cairo is for.  Something like "When Cairo is installed, drawSvg can output PNG or other image formats in addition to SVG."

cduck commented on README.md:
```diff
@@ -6,6 +6,23 @@ Most common SVG tags are supported and others can easily be added by writing a s
+Cairo needs to be installed separately. See platform-specific [instructions for Linux, Windows, and macOS from Cairo.](https://www.cairographics.org/download/) Below are some examples for installing Cairo on Linux distributions and macOS.
+
+
+
```
Nit: Remove extra blank lines.

cduck commented on drawSvg/raster.py:
```diff
@@ -1,11 +1,16 @@
-    import warnings
-    from .missing import MissingModule
+except OSError:
+    msg = 'Cairo will need to be installed to rasterize images: See prerequisites section in `README`'
```
It may not be obvious to the user where "README" is.  Link to https://github.com/cduck/drawSvg#prerequisites or https://pypi.org/project/drawSvg/ instead.

bonicim replied:
Good call. I'm going to use the anchored link to prerequisites.

cduck commented on drawSvg/raster.py:
```diff
@@ -1,11 +1,16 @@
+except OSError:
+    msg = 'Cairo will need to be installed to rasterize images: See prerequisites section in `README`'
+    cairosvg = MissingModule(msg)
+    warnings.warn(msg, RuntimeWarning)
```
Optional suggestion: Instead of catching any `OSError`, check the error message to see if it's an error about loading Cairo.

bonicim replied:
Checking the error message will prevent future misleading error messages. I'll set `msg` to be the error message of `OSError` and change `msg` only if it's an error about loading Cairo.

cduck commented on README.md:
````diff
@@ -6,6 +6,23 @@ Most common SVG tags are supported and others can easily be added by writing a s
+```
+brew install cairo
+```
+
````
It may make more sense for this to be a subsection of Install.

bonicim replied:
Good point. Will move.

bonicim commented:
Addressed feedback from @cduck. Re-tested error messages on iPython. See below:

## Tests

* Cairo not installed
```
In [1]: import drawSvg
/Users/markbonicillo/Workplace/drawSvg/drawSvg/raster.py:14: RuntimeWarning: Cairo will need to be installed to rasterize images: See prerequisites section in `README` at https://github.com/cduck/drawSvg#prerequisites
  warnings.warn(msg, RuntimeWarning)
```

* Cairo installed
```
In [1]: import drawSvg

```

cduck commented on README.md:
```diff
@@ -7,12 +7,31 @@ Most common SVG tags are supported and others can easily be added by writing a s
 
+## Prerequisites
+
+Cairo needs to be installed separately. When Cairo is installed, drawSvg can output PNG or other image formats in addition to SVG. See platform-specific [instructions for Linux, Windows, and macOS from Cairo.](https://www.cairographics.org/download/) Below are some examples for installing Cairo on Linux distributions and macOS.
```
```suggestion
Cairo needs to be installed separately. When Cairo is installed, drawSvg can output PNG or other image formats in addition to SVG. See platform-specific [instructions for Linux, Windows, and macOS from Cairo](https://www.cairographics.org/download/). Below are some examples for installing Cairo on Linux distributions and macOS.
```

cduck commented on drawSvg/raster.py:
```diff
@@ -1,11 +1,18 @@
-    from .missing import MissingModule
+except OSError as e:
+    msg = e.args[0]
+    if msg.startswith("no library called \"cairo\" was found"):
```
```suggestion
    if msg.startswith('no library called "cairo" was found'):
```

The error message reported in #4 was "OSError: dlopen() failed to load a library: cairo / cairo-2 / cairo-gobject-2".  Are there different error messages?

bonicim replied:
I reread the error message in #4 and it looks like that that person was using an Anaconda environment to load drawSvg. When I tested my error message, I used iPython and printed out the error message:

```
no library called "cairo" was found
no library called "libcairo-2" was found
cannot load library 'libcairo.so': dlopen(libcairo.so, 2): image not found
cannot load library 'libcairo.2.dylib': dlopen(libcairo.2.dylib, 2): image not found
cannot load library 'libcairo-2.dll': dlopen(libcairo-2.dll, 2): image not found
```

I initially was hesitant to use `startswith` to catch the cairo error message because it is too restrictive in it search. But I decided to use `startswith` instead of `find` because find seemed too expensive an operation to determine if an error message is related to Cairo. 

But given your question and given that drawSvg is used on multiple platforms and that there appears to be no guarantee that a Cairo-related error message will start with "no library called cairo was found", I suggest using `find` to capture a cairo-related message:

* error message from iPython
```
In [10]: msg
Out[10]: 'no library called "cairo" was found no library called "libcairo-2" was found cannot load library \'libcairo.so\': dlopen(libcairo.so, 2): image not found cannot load library \'libcairo.2.dylib\': dlopen(libcairo.2.dylib, 2): image not found cannot load library \'libcairo-2.dll\': dlopen(libcairo-2.dll, 2): image not found'

In [11]: msg.find('cairo')
Out[11]: 19
```

* error message from #4 
```
In [15]: msg = "OSError: dlopen() failed to load a library: cairo / cairo-2 / cairo-gobject-2"

In [16]: msg.find('cairo')
Out[16]: 44
```

What are your thoughts?

cduck replied:
Now that I've thought about it, I think checking the content of the message is a bad idea.  It would be more robust to print the same warning for any OSError but also include the original error message.  Something like
```
msg = ('Failed to import CairoSVG.  '
       'drawSvg will be unable to output PNG or other raster image formats.  '
       'See https://github.com/cduck/drawSvg#prerequisites for more details.\n'
       'Original OSError: {}'.format(e))
```

bonicim replied:
Agree. Including the original error message is a good idea and will help prevent masking the root cause of an OSError. Will change.

cduck commented on drawSvg/raster.py:
```diff
@@ -1,11 +1,18 @@
+except OSError as e:
+    msg = e.args[0]
+    if msg.startswith("no library called \"cairo\" was found"):
+        msg = 'Cairo will need to be installed to rasterize images: See prerequisites section in `README` at https://github.com/cduck/drawSvg#prerequisites'
```
I wouldn't even mention "README".
```suggestion
        msg = 'Cairo will need to be installed to rasterize images: See https://github.com/cduck/drawSvg#prerequisites'
```

bonicim replied:
Will change.

bonicim commented:
Retested the error message. Used [`f-strings`](https://realpython.com/python-f-strings/#simple-syntax) to format variable for the error message.  

From iPython:

```
In [1]: import drawSvg
/Users/markbonicillo/Workplace/drawSvg/drawSvg/raster.py:22: RuntimeWarning: Failed to import CairoSVG. drawSvg will be unable to output PNG or other raster image formats. See https://github.com/cduck/drawSvg#prerequisites for more details.
Original OSError: no library called "cairo" was found
no library called "libcairo-2" was found
cannot load library 'libcairo.so': dlopen(libcairo.so, 2): image not found
cannot load library 'libcairo.2.dylib': dlopen(libcairo.2.dylib, 2): image not found
cannot load library 'libcairo-2.dll': dlopen(libcairo-2.dll, 2): image not found
  warnings.warn(msg, RuntimeWarning)
  
```

cduck commented on drawSvg/raster.py:
```diff
@@ -1,11 +1,21 @@
+        'Failed to import CairoSVG. '
+        'drawSvg will be unable to output PNG or other raster image formats. '
+        'See https://github.com/cduck/drawSvg#prerequisites for more details.\n'
+        f'Original OSError: {e}'
```
Please don't use f-strings.  They require Python >=3.6.

bonicim replied:
Oh, yes. Good point. I'll use the `format`.

cduck commented on drawSvg/raster.py:
```diff
@@ -1,11 +1,21 @@
+    cairosvg = MissingModule(msg)
+    warnings.warn(msg, RuntimeWarning)
+except ImportError:
     msg = 'CairoSVG will need to be installed to rasterize images: Install with `pip3 install cairosvg`'
```
Do you want to add a similar error message here?

bonicim replied:
Yes. For completeness and to avoid masking original errors, I'll include the original import error and should be something like:


```
  msg = (
        'CairoSVG will need to be installed to rasterize images: Install with `pip3 install cairosvg`\n'
        'Original ImportError: {}'.format(e)
    )
```

bonicim commented:
Re-tested both ImportError and OSError on iPython:

```
In [1]: import drawSvg                                                                                                                                                                 
/Users/markbonicillo/Workplace/drawSvg/drawSvg/raster.py:17: RuntimeWarning: Failed to import CairoSVG. drawSvg will be unable to output PNG or other raster image formats. See https://github.com/cduck/drawSvg#prerequisites for more details.
Original OSError: no library called "cairo" was found
no library called "libcairo-2" was found
cannot load library 'libcairo.so': dlopen(libcairo.so, 2): image not found
cannot load library 'libcairo.2.dylib': dlopen(libcairo.2.dylib, 2): image not found
cannot load library 'libcairo-2.dll': dlopen(libcairo-2.dll, 2): image not found
  warnings.warn(msg, RuntimeWarning)
  
 
In [1]: import drawSvg                                                                                                                 
/Users/markbonicillo/Workplace/drawSvg/drawSvg/raster.py:24: RuntimeWarning: CairoSVG will need to be installed to rasterize images: Install with `pip3 install cairosvg`
Original ImportError: No module named 'cairosvg'
  warnings.warn(msg, RuntimeWarning)
```

cduck commented:
Published version 1.3.1 on PyPI with this fix: https://pypi.org/project/drawSvg/1.3.1/
